sycl/plugins/cuda/pi_cuda.cpp

```diff
@@ -1909,23 +1909,90 @@ pi_result cuda_piProgramGetInfo(pi_program program, pi_program_info param_name,
   return {};
 }
 
-pi_result cuda_piProgramLink( // TODO: change interface to return error code
+pi_result cuda_piProgramLink(
     pi_context context, pi_uint32 num_devices, const pi_device *device_list,
     const char *options, pi_uint32 num_input_programs,
     const pi_program *input_programs,
     void (*pfn_notify)(pi_program program, void *user_data), void *user_data,
     pi_program *ret_program) {
-  cl::sycl::detail::pi::die("cuda_piProgramLink not implemented");
-  return {};
+  
+  assert(ret_program != nullptr);
+  assert(num_devices == 1 || num_devices == 0);
+  assert(device_list != nullptr || num_devices == 0);
+  assert(pfn_notify == nullptr);
+  assert(user_data == nullptr);
+  pi_result retError = PI_SUCCESS;
+
+  try {
+    ScopedContext active(context);
+
+    CUlinkState state;
+    std::unique_ptr<_pi_program> retProgram{new _pi_program{context}};
+
+    // TODO: Linker options
+    retError = PI_CHECK_ERROR(cuLinkCreate(0, nullptr, nullptr, &state));
+    try {
+      for (size_t i; i < num_input_programs; ++i) {
+        pi_program program = input_programs[i];
+        std::vector<char> source(program->source_,
+                                program->source_ + program->sourceLength_);
+        retError = PI_CHECK_ERROR(cuLinkAddData(
+            state, CU_JIT_INPUT_PTX, source.data(), program->sourceLength_,
+            nullptr, 0, nullptr, nullptr));
+      }
+      void *cubin = nullptr;
+      size_t cubinSize = 0;
+      retError = PI_CHECK_ERROR(cuLinkComplete(state, &cubin, &cubinSize));
+
+      retError = retProgram->create_from_source(
+          static_cast<const char *>(cubin), cubinSize);
+
+      if (retError != PI_SUCCESS) {
+        return retError;
+      }
+
+      retError = retProgram->build_program(options);
```

Comment:
I'm not really familiar with the cuda, so, probably silly question: why do you at first link something, then create program from sources (which are actually binaries from link stage) and then build it?

Reply:
According to the CUDA Driver API documentation;
> Completes the pending linker action and returns the cubin image for the linked device code, which can be used with `cuModuleLoadData`.

So once we've completed the linking we have to load the module we're given. To do this we set the module data of `retProgram` using `create_from_source` and instruct it to load the module through the `build_program` call.

Reply:
Ok, thanks